src/jekyll/jk_document.rs

```diff
@@ -0,0 +1,285 @@
+        let mut custom_attributes = jk_front.0;
+        frontmatter::FrontmatterBuilder::new()
+            .merge_slug(custom_attributes
+                            .remove("slug")
+                            .and_then(|v| v.as_str().map(|s| s.to_owned()))
+                            .or_else(|| {
+                                         Some(slugify(custom_attributes
+                                 .get("title")
+                                 .unwrap_or(&liquid::Value::str("No Title"))
+                                 .as_str()
+                                 .unwrap()))
+                                     })
+                            .map(|s| s.to_owned()))
+            .merge_title(custom_attributes
+                             .remove("title")
+                             .and_then(|v| v.as_str().map(|s| s.to_owned())))
+            .merge_description(custom_attributes
+                                   .remove("excerpt")
+                                   .and_then(|v| v.as_str().map(|s| s.to_owned())))
+            .merge_categories(custom_attributes
+                                  .remove("categories")
+                                  .and_then(|v| {
+                                                v.as_array()
+                                                    .map(|v| {
+                                                             v.iter()
+                                                                 .map(|v| v.to_string())
+                                                                 .collect()
+                                                         })
+                                            }))
+            .merge_permalink(custom_attributes
+                                 .remove("permalink")
+                                 .and_then(|v| v.as_str().map(|s| s.to_owned())))
+            .merge_draft(custom_attributes
+                             .remove("published")
+                             .and_then(|v| v.as_bool().map(|b| !b)))
+            .merge_layout(custom_attributes
+                              .remove("layout")
+                              .and_then(|v| v.as_str().map(|s| s.to_owned())))
+            .merge_published_date(custom_attributes
+                                      .remove("date")
+                                      .and_then(|d| {
+                                                    d.as_str().and_then(datetime::DateTime::parse)
+                                                }))
+            .merge_custom(custom_attributes)
+    }
+}
+
+#[derive(Debug, Default, Clone, Eq, PartialEq, Serialize, Deserialize)]
+pub struct JkDocument {
+    pub front: Option<String>,
+    pub content: Option<String>,
+}
+
+impl JkDocument {
+    pub fn parse_string(doc: String) -> Result<JkDocument> {
+        let (front, content) = split_document(&doc)?;
+        if front != None {
+            Err(ErrorKind::MissingFrontStart.into())
```

Comment:
`ok_or_else` will be a help here

```rust
let (front, content) = split_document(&doc)?;
let front = front.ok_or_else(|| Err(ErrorKind::MissingFrontStart.into())?;
let (front, content) = split_document(content)?;
let front = front.ok_or_else(|| Err(ErrorKind::MissingFrontStart.into())?;
Ok(JkDocument {front: Some(front.to_owned()), content: Some(content.to_owned()),})
```

See also https://rustbyexample.com/error/multiple_error_types/define_error_type.html